pkg/sql/crdb_internal.go

```diff
@@ -218,86 +218,93 @@ CREATE TABLE crdb_internal.tables (
   audit_mode               STRING NOT NULL,
   schema_name              STRING NOT NULL
 )`,
-	populate: func(ctx context.Context, p *planner, _ *DatabaseDescriptor, addRow func(...tree.Datum) error) error {
-		descs, err := p.Tables().getAllDescriptors(ctx, p.txn)
-		if err != nil {
-			return err
-		}
-		dbNames := make(map[sqlbase.ID]string)
-		// Record database descriptors for name lookups.
-		for _, desc := range descs {
-			db, ok := desc.(*sqlbase.DatabaseDescriptor)
-			if ok {
-				dbNames[db.ID] = db.Name
+	generator: func(ctx context.Context, p *planner, dbDesc *DatabaseDescriptor) (virtualTableGenerator, func(), error) {
+		row := make(tree.Datums, 14)
+		worker := func(addRow func(...tree.Datum) error) error {
+			descs, err := p.Tables().getAllDescriptors(ctx, p.txn)
+			if err != nil {
+				return err
 			}
-		}
-
-		addDesc := func(table *sqlbase.TableDescriptor, dbName tree.Datum, scName string) error {
-			leaseNodeDatum := tree.DNull
-			leaseExpDatum := tree.DNull
-			if table.Lease != nil {
-				leaseNodeDatum = tree.NewDInt(tree.DInt(int64(table.Lease.NodeID)))
-				leaseExpDatum = tree.MakeDTimestamp(
-					timeutil.Unix(0, table.Lease.ExpirationTime), time.Nanosecond,
-				)
+			dbNames := make(map[sqlbase.ID]string)
+			// Record database descriptors for name lookups.
+			for _, desc := range descs {
+				db, ok := desc.(*sqlbase.DatabaseDescriptor)
+				if ok {
+					dbNames[db.ID] = db.Name
+				}
 			}
-			dropTimeDatum := tree.DNull
-			if table.DropTime != 0 {
-				dropTimeDatum = tree.MakeDTimestamp(
-					timeutil.Unix(0, table.DropTime), time.Nanosecond,
+
+			addDesc := func(table *sqlbase.TableDescriptor, dbName tree.Datum, scName string) error {
+				leaseNodeDatum := tree.DNull
+				leaseExpDatum := tree.DNull
+				if table.Lease != nil {
+					leaseNodeDatum = tree.NewDInt(tree.DInt(int64(table.Lease.NodeID)))
+					leaseExpDatum = tree.MakeDTimestamp(
+						timeutil.Unix(0, table.Lease.ExpirationTime), time.Nanosecond,
+					)
+				}
+				dropTimeDatum := tree.DNull
+				if table.DropTime != 0 {
+					dropTimeDatum = tree.MakeDTimestamp(
+						timeutil.Unix(0, table.DropTime), time.Nanosecond,
+					)
+				}
+				row = row[:0]
```

Comment:
i'm wondering if this is dangerous. do rows get copied else? can we add documentation that "assumes rows from addRow have been copied out and is safe to reuse"

Reply:
They get copied as part of the next interface if needed, but it makes sense to add this to the setupGenereator docs.